Probability Metric + New Normalization

hynky1999 opened the pull request:
What does this implement/fix? Explain your changes.
---------------------------------------------------
This PR adds two new features:
1) New Probability Metric, allowing to collect probability of correct answer. This can be either raw prob or prob mass (normalized by other choices)
2) Revamps Acc/Prob normalization and adds two new normalizations 
      a) Token normalization, which we found to be better at most of the non-english langauges compared to acc norm.
      b) PointwiseMutualInformation normalization, which is good way for testing tasks with unlikely token see: https://arxiv.org/abs/2406.08446

Lastly I have done some small changes to the requests processing, removing parts, which are not needed and can easily cause bugs.


Comments
----------
- I am not really content with having new category just for normalization but I didn't find a better way in the current system. The problem is that when creating requests we only have access to sample fc, but nothing else, thus we can't really do any kind of structural decomposition :(
- This new norms are only added for non-single token types of tasks. Adding them to single token would require improving the requests creating logic to be maintanable and can be done in other PR  

PS: Relevant disscusion about token norm https://github.com/EleutherAI/lm-evaluation-harness/issues/1396

clefourrier commented on src/lighteval/metrics/normalizations.py:
```diff
@@ -349,3 +350,80 @@ def gsm8k_normalizer(text: str) -> str:
+Normalization = CharNorm | TokenNorm | PMINorm
+
+
+def normalize_log_probs(
```
Why don't you use the mechanism we have for the metrics with arguments, where you instantiate a class then execute the correct normalisation based on the arguments?
Using empty dataclasses, then case and asserts has a very different style from the rest of the lib

clefourrier replied:
```
class Normalization:
    def __init__(self, level: str, ignore_first_space: bool = False):
        if self.level not in ["character", "token", "pmi"]:
            raise ValueError(...)
        self.level = level
        self.ignore_first_space = ignore_first_space
        
    def normalize(
        self, 
        choices_logprob: list[float],
        unconditioned_logprob: list[float] | None,
        choices_text: list[str] | None,
        choices_tokens: list[list[int]] | None,
    ): ...
```

clefourrier replied:
Btw normalization is too broad as a name - you'll need to find something else

hynky1999 replied:
Because ignore_first_space is property of just Char_normalization. The parameter doesn't make sense in other normalization context so it make sense to me to be a property of Character normalization

clefourrier replied:
Then I would be OK with one class for each - I just don't see the points of the empty dataclasses which add an extra layer of complexity to the code

hynky1999 replied:
So at the end your proposed change would be to switch to literals (by flattening, so that char norm has now 2 literals), but would keep the rest of the logic the same ?

Like in the example you shown I don't really see why we need any class at all, we don't need to keep any state there or whatsover we just need to know how to normalize

clefourrier replied:
> I don't really see why we need any class at all, we don't need to keep any state there or whatsover we just need to know how to normalize
1) homegeneity with the rest of the code style
2) avoiding having many functions (normalisation, to string), which are unconnected with the code base.
Give me a minute i'll send you how I would write this

hynky1999 replied:
I can also do it on OOP way.
```
class Normalization:
    def normalize(logprobs):
          pass
          
          
class CharNorm(Normalization):
      def normalize(logprobs):
          pass
```

Probably better as we will rather be adding new Normalizations than adding new functions that use them. wdyt @clefourrier ?

NathanHB replied:
Not a fan of the OOP way, the different norms having nothing in common so having the hinerit from a common class is weird at best.
The way you did it here is the best I think but clementine is right in that it does not match the rest of the codebase, I would however still go for that approach.

clefourrier commented on src/lighteval/tasks/requests.py:
```diff
@@ -177,6 +177,9 @@ class Doc:
     num_asked_few_shots: int = -1
     num_effective_few_shots: int = -1
 
+    # For PMI normalization
```
This needs more doc to explain the diff with normal query

clefourrier commented on src/lighteval/tasks/lighteval_task.py:
```diff
@@ -744,7 +729,7 @@ def create_requests_from_tasks(  # noqa: C901
                     reqs = task.construct_requests(doc, ctx, doc_id_seed, cur_task_name)
-                    for req_type in req_types:
-                        requests[req_type].extend(reqs[req_type])
+                    for req_type, reqs in reqs.items():
```
overwritting `reqs` locally could lead to unexpected behavior. you could use `for req_type, req in reqs.items():` instead

hynky1999 replied:
Ahh yeah that's so stupid, good spot :thumbsup:

clefourrier commented on src/lighteval/metrics/metrics.py:
```diff
@@ -644,3 +650,45 @@ def all_metrics():
         return res
+
+    @staticmethod
+    def loglikelihood_acc_metric(normalization: Normalization | None = None) -> SampleLevelMetric:
+        """
+        Function for creating custom loglikelihood accuracy metric in runtime.
+        """
+
+        normalization_str = normalization_name(normalization) if normalization else ""
+        metric_name = f"acc_{normalization_str}"
+        return SampleLevelMetric(
+            metric_name=metric_name,
+            sample_level_fn=LoglikelihoodAcc(normalization=normalization).compute,
+            category=MetricCategory.MULTICHOICE if not normalization == PMINorm() else MetricCategory.MULTICHOICE_PMI,
+            use_case=MetricUseCase.ACCURACY,
+            corpus_level_fn=np.mean,
+            higher_is_better=True,
+        )
+
+    @staticmethod
+    def probability_metric(
+        normalization: Normalization | None = None,
+        return_mass: bool = False,
+        aggregation_function: Callable[[np.ndarray], float] = np.max,
+    ) -> SampleLevelMetric:
+        """
+        Function for creating custom probability metric in runtime.
+        """
+
+        mass_str = "mass" if return_mass else ""
+        normalization_str = normalization_name(normalization) if normalization else ""
+        metric_name = "_".join(filter(None, ["prob", mass_str, normalization_str]))
+
+        return SampleLevelMetric(
+            metric_name=metric_name,
+            sample_level_fn=Probability(
```
This does not go here - it has nothing to do in the enum class, and it's not even needed since these functions are static

hynky1999 replied:
Wdym by other dtailed metrics files ?
I too wasn't sure where to put this, but essentially there are full fledged metrics (just need few args to initialize), I don't something like this already in the codebase.
I placed them here because indeed the enum already have few staticmethods.

ANyways I guess I can create a new file dynamic|partial_metrics.py wdyt ?

clefourrier replied:
dynamic_metrics works for me!

clefourrier commented on src/lighteval/metrics/metrics_sample.py:
```diff
@@ -207,47 +207,114 @@ def compute_one_item(self, gold: str, pred: str) -> float:
+        self.aggregation_function = aggregation_function
+        self.return_mass = return_mass
+
+    def safe_divide(self, numerator: np.ndarray, denominator: float, default_value: float = 0.0) -> np.ndarray:
+        return np.where(denominator != 0, numerator / denominator, default_value)
```
could go in lighteval.utils.utils

clefourrier commented on src/lighteval/metrics/normalizations.py:
```diff
@@ -349,3 +350,80 @@ def gsm8k_normalizer(text: str) -> str:
+    return normalized_log_probs
+
+
+def normalization_name(normalization: Normalization) -> str:
```
same, this could be replaced by a __str__ in the above class

clefourrier commented on src/lighteval/tasks/lighteval_task.py:
```diff
@@ -517,6 +483,22 @@ def construct_requests(
             ]
+
+        if self.has_metric_category[MetricCategory.MULTICHOICE_PMI]:
+            assert (
```
Why do you add Multichoice_PMI 2 times ? Do you need to create 2 requests ?

hynky1999 replied:
I do need to create two requests one with the `uncoditioned_query` and one with normal query

clefourrier replied:
OK!

clefourrier commented on tests/metrics/test_normalizations.py:
```diff
@@ -0,0 +1,63 @@
+# MIT License
```
thanks for adding tests!

clefourrier commented on tests/test_unit_base_metrics.py:
```diff
@@ -180,6 +182,103 @@ def test_prefix_quasi_exact_match(self):
         res = em.compute_one_item("", "")
         assert res == 0
 
+    def test_prob(self):
```
Nice, thanks!

hynky1999 commented:
> I'm still not clear why you need a new system for the PMI requests, can you elaborate?

Because PMI needs two requests, one with normal query and with with uncoditioned query (Usually empty or just Answer:).
Thus you need to create two loglikehood requests

clefourrier commented on src/lighteval/logging/info_loggers.py:
```diff
@@ -207,6 +207,7 @@ class Detail:
         gold_index: list = field(default_factory=list)
         metrics: dict = field(default_factory=dict)
         specifics: dict = field(default_factory=dict)
+        unconditioned_query: Optional[str] = None
```
This should go into specifics instead - that's where we put all the task/metric specific added logic

NathanHB commented on src/lighteval/logging/info_loggers.py:
```diff
@@ -369,6 +369,12 @@ def log(
             detail.choices = doc.choices
             detail.gold_index = as_list(doc.gold_index)
             pred_saved = True
+
```
```suggestion
```

NathanHB commented on src/lighteval/logging/info_loggers.py:
```diff
@@ -451,6 +456,7 @@ def aggregate(self):
         )
 
 
+@dataclass
```
this should have been merged in #280

NathanHB commented on src/lighteval/metrics/__init__.py:
```diff
@@ -124,23 +124,44 @@ def apply_generative_metric(
 
+    if is_pmi_category and len(results) != n_choices * 2:
+        raise Exception(
+            f"You shoud have returned as many model outputs as choices when using an multi choice metric. Returned {len(results)} instead of {n_choices * 2} (conditioned + unconditioned)"
```
```suggestion
            f"You shoud have returned twice as many model outputs as choices when using an probability multi choice metric. Returned {len(results)} instead of {n_choices * 2} (conditioned and unconditioned)"
```

NathanHB commented on src/lighteval/metrics/dynamic_metrics.py:
```diff
@@ -0,0 +1,71 @@
+from lighteval.metrics.utils import MetricCategory, MetricUseCase, SampleLevelMetric
+
+
+def loglikelihood_acc_metric(normalization: Normalization | None = None) -> SampleLevelMetric:
```
The be coherent with the rest of the codebase

```suggestion
def loglikelihood_acc_metric(normalization: Optional[Normalization] = None) -> SampleLevelMetric:
```

NathanHB replied:
will need to add the import statement

NathanHB replied:
OK so after looking it up, the prefered way is indeed using `type | None`, ^^'

NathanHB commented on src/lighteval/metrics/harness_compatibility/truthful_qa.py:
```diff
@@ -22,9 +22,17 @@
+def truthfulqa_mc_metrics(
+    gold_ixs: list[int],
+    choices_logprob: list[float],
+    unconditioned_logprob: list[float] | None,
```
```suggestion
    unconditioned_logprob: Optional[list[float]],
```

NathanHB commented on src/lighteval/metrics/harness_compatibility/truthful_qa.py:
```diff
@@ -22,9 +22,17 @@
+    gold_ixs: list[int],
+    choices_logprob: list[float],
+    unconditioned_logprob: list[float] | None,
+    choices_tokens: list[list[int]] | None,
```
```suggestion
    choices_tokens: Optional[list[list[int]]],
```

NathanHB commented on src/lighteval/metrics/metrics_sample.py:
```diff
@@ -207,47 +207,113 @@ def compute_one_item(self, gold: str, pred: str) -> float:
 
 class LoglikelihoodAcc:
-    def __init__(self, length_normalization: bool = False, ignore_first_space: bool = False) -> None:
+    def __init__(self, normalization: Normalization | None = None):
```
```suggestion
    def __init__(self, normalization: Optional[Normalization] = None):
```

NathanHB commented on src/lighteval/metrics/metrics_sample.py:
```diff
@@ -207,47 +207,113 @@ def compute_one_item(self, gold: str, pred: str) -> float:
+        self,
+        gold_ixs: list[int],
+        choices_logprob: list[float],
+        unconditioned_logprob: list[float] | None,
```
```suggestion
        unconditioned_logprob: Optional[list[float]],
```

NathanHB commented on src/lighteval/metrics/metrics_sample.py:
```diff
@@ -207,47 +207,113 @@ def compute_one_item(self, gold: str, pred: str) -> float:
+        gold_ixs: list[int],
+        choices_logprob: list[float],
+        unconditioned_logprob: list[float] | None,
+        choices_tokens: list[list[int]] | None,
```
```suggestion
        choices_tokens: Optional[list[list[int]]],
```

NathanHB commented on src/lighteval/metrics/metrics_sample.py:
```diff
@@ -207,47 +207,113 @@ def compute_one_item(self, gold: str, pred: str) -> float:
+class Probability:
+    def __init__(
+        self,
+        normalization: Normalization | None = None,
```
```suggestion
        normalization: Optional[Normalization] = None,
```

NathanHB commented on src/lighteval/metrics/metrics_sample.py:
```diff
@@ -207,47 +207,113 @@ def compute_one_item(self, gold: str, pred: str) -> float:
+        self,
+        gold_ixs: list[int],
+        choices_logprob: list[float],
+        unconditioned_logprob: list[float] | None,
```
```suggestion
        unconditioned_logprob: Optional[list[float]],
```

NathanHB commented on src/lighteval/metrics/metrics_sample.py:
```diff
@@ -207,47 +207,113 @@ def compute_one_item(self, gold: str, pred: str) -> float:
+        gold_ixs: list[int],
+        choices_logprob: list[float],
+        unconditioned_logprob: list[float] | None,
+        choices_tokens: list[list[int]] | None,
```
```suggestion
        choices_tokens: Optional[list[list[int]]],
```

NathanHB commented on src/lighteval/tasks/requests.py:
```diff
@@ -184,6 +184,11 @@ class Doc:
 
+    # Uncoditioned query is used for PMI normalization, that's
+    # log P(choice | Query) - log P(choice | Unconditioned Query)
+    # The uncoditiioned query shouldn't contain any information about the task, thus usually it's empty string or 'Answer:'.
```
```suggestion
    # The uncoditioned query shouldn't contain any information about the task, thus usually it's empty string or 'Answer:'.
```

NathanHB commented on tests/metrics/test_metric_requests.py:
```diff
@@ -0,0 +1,178 @@
+    pmi_test_config = get_pmi_task(metrics=metrics)
+    task = LightevalTask(pmi_test_config.name, pmi_test_config)
+    result = fake_evaluate_task(task, fake_model, max_samples=1)
+    # Correc choice after norm should be the second one so 0 acc
```
```suggestion
    # Correct choice after norm should be the second one so 0 acc
```

hynky1999 commented:
Rebase and made a small update:
- Moved the raw probability to target_perplexity category to reduce wasteful compute
- This meant that I had to change the singature that the comp fcs are receiving, so I updated that

Also noticed that we have this metric `acc_gold_likelihoods`, but it was computing completely different thing than expected so updated to to cmpute what it should:

Here are comments I wrote to that fc
```
    # TODO: (hynek) I think this function is absolutely broken and doesn't work as intended. The target_accs used to be True/False
    # True if argmax of logits is in the gold, False otherwise. We then took only the first gold for some reason.

    # Thus the name doesn't reflect that at all and neither do comments
    # 1) I think you wanted to talk about probs
    # 2) Even that's not true, what this measure is whether every prob of token will be > than the others
    # 3) This was measure across just single gold, not across all possible golds
 ```

NathanHB commented on src/lighteval/metrics/sample_preparator.py:
```diff
@@ -138,4 +138,8 @@ def prepare(self, logprobs: list[float] | float, reference_text: str, **kwargs):
         """
-        return PerplexityCorpusMetricInput(logprobs=logprobs, weights=self.count_units(reference_text))
+
+        print(logprobs)
```
```suggestion
```

NathanHB commented on src/lighteval/tasks/lighteval_task.py:
```diff
@@ -513,7 +515,7 @@ def get_metric_method_from_category(self, metric_category):
         if metric_category == MetricCategory.TARGET_PERPLEXITY:
             return apply_target_perplexity_metric
-        if metric_category == MetricCategory.MULTICHOICE:
+        if metric_category == MetricCategory.MULTICHOICE or metric_category == MetricCategory.MULTICHOICE_PMI:
```
```suggestion
        if metric_category in [MetricCategory.MULTICHOICE, MetricCategory.MULTICHOICE_PMI]:
```

NathanHB commented:
Thanks for the modifs and the tests ! This looks good to be merged after running `ruff`. For the normalization functions, I would go for the way you use a function to route to the correct behaviour as we do not need OOP here.

hynky1999 commented:
Ruffed, let's wait for tests and merge it
